wurst/systems/spells/mage/dementia/DarkGateNew.wurst

```diff
@@ -0,0 +1,405 @@
+                    orb.castRandomSpell(t)
+                    _foundTarget = true
+            
+
+    function updateOrbAngles()
+        //Calculate orb angles and set it
+        angle orbOffset = angle((360 / orbs.size())*DEGTORAD)
+        int i = 0
+        for o in orbs
+            angle indexOrbAngle = orbAngle + orbOffset * i.toReal()
+            o.setPos(centerPos.offset2d(indexOrbAngle, portalRadius))
+            o.setFacing(o.pos.angleTo2d(centerPos))
+            i = i + 1
+
+    function startPortalLoop()
+        if channeledTime < MINIMUM_GROW_DURATION
+            destroy this
+        else
+            portalActiveLoop = doPeriodicallyTimed(ORB_LOOP_INTERVAL, durationBuildup) portalLoop ->
+                //Check for destruction conditions first before accessing any objects
+                if (portalLoop.isLast())
+                    destroy this
+                else    
+                    portalLoopUpdate()
+        
+    function portalLoopUpdate()
+        //Check time since last spell was cast and cast
+        if (this.elapsedLifetime - this.lastSpellCastTimestamp >= this.spellIntervalCurrent)
+            this.lastSpellCastTimestamp = this.elapsedLifetime
+            randomOrbCastSpell()
+        elapsedLifetime += ORB_LOOP_INTERVAL
+        orbAngle = angle(orbAngle.radians() + ORB_LOOP_INTERVAL * ORB_ANGULARSPEED_ACTIVE * DEGTORAD)
+        updateOrbAngles()
+        updateSummons()
+
+    function isValidEnemyTarget(unit t) returns boolean
+        return t.getOwner().isEnemyOf(ownerUnit.getOwner()) and not t.isInvulnerable() and not t.isType(UNIT_TYPE_STRUCTURE) and t.isAlive() and IsUnitVisible(t, ownerUnit.getOwner())
+
+    function isValidAllyTarget(unit t) returns boolean
+        return not t.isInvulnerable() and not t.isType(UNIT_TYPE_STRUCTURE) and t.isAlive() and IsUnitVisible(t, ownerUnit.getOwner())
+
+    function summonTryAttacking(Summon s)
+        //Checking time since last attack
+        if (s.elapsedLifetime - s.lastAttack >= s.cooldown)
+
+            //Create object for closure to store unit
+            unitObject _summonTarget = new unitObject(null)
+            forUnitsInRange(centerPos.toVec2(), portalRadius) t ->
+                //Try to find enemies first, pick targets without buffs over those who have em
+                if _summonTarget.value == null or _summonTarget.value.hasAbility('Bmlt') or _summonTarget.value.hasAbility('B00Q')
+                    if isValidEnemyTarget(t)
+                        _summonTarget.value = t
+
+                if _summonTarget.value == null or _summonTarget.value.hasAbility('Bmlt') or _summonTarget.value.hasAbility('B00Q')
+                    if isValidAllyTarget(t)
+                        _summonTarget.value = t
+                
+            if _summonTarget.value != null
+                s.summonUnit.issueTargetOrderById(s.orderId, _summonTarget.value)
+                s.lastAttack = s.elapsedLifetime
+
+            destroy _summonTarget
```

Comment:
Perhaps we can do something like

```
let pos = centerPos.toVec2()
forNearestUnit(pos, portalRadius, Filter(-> begin
    let filter = GetFilterUnit()
    return not filter.hasAbility('Bmlt') and not filter.hasAbility('B00Q') and isValidEnemyTarget(filter)
end) nearest ->
    if nearest != null
        s.summonUnit.issueTargetOrderById(s.orderId, nearest)
    else
        forNearestUnit(pos, portalRadius, Filter(-> isValidEnemyTarget(GetFilterUnit()))) nearest ->
            if nearest != null
                s.summonUnit.issueTargetOrderById(s.orderId, nearest)
```

Not a huge fan of the double `forNearestUnit` but it will work. Alternatively, we can do something like this:

```
import GroupExtensions

var targets = CreateGroup()
    ..enumUnitsInRange(pos, range, Filter(-> isValidTarget(GetFilterUnit()))
    .toLinkedList()

var without = targets.filter(u -> not u.hasAbility('Bmlt') and not u.hasAbility('B00Q'))
if without.size() > 0
    destroy targets
    targets = without

let target = targets
    ..sortBy((a, b) -> pos.distanceToSq(a.getPos()) - pos.distanceToSq(b.getPos()))
    .getFirst()

   s.summonUnit.issueTargetOrderById(s.orderId, target)

destroy targets
```

Added benefit of these options is that it provides an additional sort order on the units--their distance. If you dislike both these options, it is totally fine to stick with your current code. If you do, I would ask that you rename `unitObject` to `UnitObject` and I think you have duplicate code on lines 325-327. Happy to discuss further in Discord.

Reply:
The last option looks cleaner, the only thing is that the hasAbility checks are optional. So if there is a target with a buff we CAN attack that, but we prioritize targets without buffs. Can we do another sort where the other comparison function just returns 0 for any unit with no buffs and 1 unit if it has buffs?

Reply:
added